counsel.el

```diff
@@ -2063,6 +2063,65 @@ By default `counsel-bookmark' opens a dired buffer for directories."
    ("r" ,(counsel--apply-bookmark-fn #'counsel-find-file-as-root)
         "open as root")))
 
+;;** `counsel-bookmarked-directory'
+(defcustom counsel-bookmarked-directory-default-name "%s"
+  "Default bookmark name created by `counsel-bookmarked-directory-add."
+  :group 'ivy
+  :type 'string)
+
+(defun counsel-bookmarked-directory--path-separator ()
+  "Get the path separator for the current system."
+  (let* ((parent (expand-file-name "a"))
+         (i (length parent)))
+    (substring (expand-file-name "b" parent) i (1+ i))))
+
+(defconst counsel-bookmarked-directory-separator
+  (counsel-bookmarked-directory--path-separator))
+
+(defun counsel-bookmarked-directory--candidates ()
+  "Get a list of bookmarked directories sorted by file path."
+  (bookmark-maybe-load-default-file)
+  (cl-sort (cl-remove-if-not
+            (lambda (filename)
+              (string-suffix-p counsel-bookmarked-directory-separator filename))
```

Comment:
Given my previous comment, this `lambda` can be replaced with `#'ivy--dirname-p`.